content/posts/observability-in-the-cloud.md

```diff
@@ -0,0 +1,86 @@
+- feature flags
+- shorter deployment cycles
+
+These changes lead to a more complex application landscape, which makes debugging more and more infeasible.
+Instead of running one service in a debugger and recreating its state, we may need to start various services and try to reproduce state in each of them.
+Our telemetry may also become less expressive and we may have to ask ourselves additional questions, such as:
+
+- has a deployment just happened?
+- what versions are actually deployed?
+- which versions (if multiple are running) show anomalous behaviour
+- which feature flags were enabled for the requests/tasks that failed?
+
+To ensure, we still get some meaningful insights from our telemetry, we need to annotate it with additional information such as build ID, hostname, IP address, enabled feature flags and more.
+It also brings about new problems: adding these dimensions to our telemetry makes the storage and querying systems much slower and/or more expensive.
+At the same time, we still need to make a number of queries to correlate information from multiple source systems (microservices) to slowly piece together enough context to guess what happened.
+
+## Closing the Observability Gap in Distributed Cloud Systems
+
+Unfortunately, this gap cannot be closed by the simple addition of a new tool or library.
+Instead, there are two primary challenges that should be tackled by site reliability engineers who aim to improve observability in their systems.
+
+First, each individual service should be instrumented to keep track of its context and emit arbitrarily wide events containing all relevant context when a request/task processing finishes (successfully or with an exception).
+To do this, whenever a workload is started (e.g. a request enters a system), an event context object should be initialized and prepopulated with all relevant data.
+As the workload is being processed, this object should be enriched and whenever another service is called, relevant context should be passed downstream.
+When the request finishes, the whole context event should be passed to the central storage system.
+Many logging libraries do support such use cases, but it often depends on the framework used.
+Ensuring, that these events are properly enriched is the responsibility of developers and site reliability engineers.
+The goal should be, that problems can usually be pinpointed by looking at a single event, without having to painstakingly correlate dozens of events or reproduce an issue through incremental trial and error.
+
+As a second challenge, an event storage and querying system must be put in place, that is able to reasonably process these wide events and allows querying and searching.
+Storing arbitrarily wide events is crucial and querying across all dimensions is crucial to avoid limitations to the sort of questions you can ask, but it is not supported by many monitoring systems, especially since it's usually very resource intensive.
```

Comment:
```suggestion
Storing arbitrarily wide events and querying across all dimensions is crucial in order to avoid limitations to the sort of questions you can ask. However, it is not supported by many monitoring systems, especially since it is usually very resource intensive.
```